strawberryfields/engine.py

```diff
@@ -531,161 +544,542 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
 
         Returns:
-            (strawberryfields.api_client.Job): a Job instance referencing the queued job
+            strawberryfields.engine.Job: the job
         """
-        job = Job(client=self.client)
-        job.manager.create(circuit=job_content)
-        self.jobs.append(job)
-        print("Job {} is sent to server.".format(job.id.value))
-        return job
+        response = self._get("/jobs/{}".format(job_id))
+        if response.status_code == 200:
+            return Job(
+                id_=response.json()["id"],
+                status=JobStatus(response.json()["status"]),
+                connection=self,
+            )
+        raise RequestFailedError(self._format_error_message(response))
+
+    def get_job_status(self, job_id: str) -> JobStatus:
+        """Returns the status of a job.
+
+        Args:
+            job_id (str): the job ID
 
-    def _run_program(self, program, **kwargs):
+        Returns:
+            strawberryfields.engine.JobStatus: the job status
         """
-        Given a compiled program, gets the blackbird circuit code and creates (or resumes) a job
-        via the API. If the job is completed, returns the job result.
+        return JobStatus(self.get_job(job_id).status)
 
-        A queued job can be interrupted by a ``KeyboardInterrupt`` event, at which point if the
-        job ID was retrieved from the server, the job will be accessible via
-        :meth:`~.Starship.jobs`.
+    def get_job_result(self, job_id: str) -> Result:
+        """Returns the result of a job.
 
         Args:
-            program (strawberryfields.program.Program): program to be executed remotely
+            job_id (str): the job ID
 
         Returns:
-            (list): a list representing the result samples
+            strawberryfields.engine.Result: the job result
+        """
+        response = self._get(
+            "/jobs/{}/result".format(job_id), {"Accept": "application/x-numpy"},
+        )
+        if response.status_code == 200:
+            # Read the numpy binary data in the payload into memory
+            with io.BytesIO() as buf:
+                buf.write(response.content)
+                buf.seek(0)
+                samples = np.load(buf)
+            return Result(samples, is_stateful=False)
+        raise RequestFailedError(self._format_error_message(response))
+
+    def cancel_job(self, job_id: str):
+        """Cancels a job.
 
-        Raises:
-            Exception: In case a job could not be submitted or completed.
-            TypeError: In case a job is already queued and a user is trying to submit a new job.
+        Args:
+            job_id (str): the job ID
         """
-        if self.jobs:
-            raise OneJobAtATimeError(
-                "A job is already queued. Please reset the engine and try again."
-            )
+        response = self._patch(
+            "/jobs/{}".format(job_id), data={"status", JobStatus.CANCELLED.value}
+        )
```

Comment:
Just for clarification: you are using `PATCH` here such that you can modify *only* the status, right?

Reply:
That's right!